modules/src/core/ics02_client/client_state.rs

```diff
@@ -1,40 +1,62 @@
+use core::fmt::{Debug, Display};
 use core::marker::{Send, Sync};
 use core::time::Duration;
 
+use derivative::Derivative;
 use ibc_proto::google::protobuf::Any;
 use serde::{Deserialize, Serialize};
 use tendermint_proto::Protobuf;
 
-use ibc_proto::ibc::core::client::v1::IdentifiedClientState;
-
+use crate::clients::ics07_tendermint::client_def::TendermintClient;
 use crate::clients::ics07_tendermint::client_state;
+use crate::clients::ics07_tendermint::consensus_state::ConsensusState as TendermintConsensusState;
 #[cfg(any(test, feature = "ics11_beefy"))]
 use crate::clients::ics11_beefy::client_state as beefy_client_state;
+#[cfg(any(test, feature = "ics11_beefy"))]
+use crate::clients::ics11_beefy::{
+    client_def::BeefyClient, consensus_state::ConsensusState as BeefyConsensusState,
+};
+use crate::clients::{ClientStateOf, ConsensusStateOf, GlobalDefs};
+use crate::core::ics02_client::client_def::{AnyClient, ClientDef};
 use crate::core::ics02_client::client_type::ClientType;
 use crate::core::ics02_client::error::Error;
 use crate::core::ics02_client::trust_threshold::TrustThreshold;
 use crate::core::ics24_host::error::ValidationError;
 use crate::core::ics24_host::identifier::{ChainId, ClientId};
 #[cfg(any(test, feature = "mocks"))]
-use crate::mock::client_state::MockClientState;
+use crate::mock::client_def::MockClient;
+#[cfg(any(test, feature = "mocks"))]
+use crate::mock::client_state::{MockClientState, MockConsensusState};
 use crate::prelude::*;
 use crate::Height;
+use ibc_proto::ibc::core::client::v1::IdentifiedClientState;
+
+#[cfg(not(feature = "ics11_beefy"))]
+use super::client_def::stub_beefy::*;
+#[cfg(not(test))]
+use super::client_def::stub_mock::*;
 
 pub const TENDERMINT_CLIENT_STATE_TYPE_URL: &str = "/ibc.lightclients.tendermint.v1.ClientState";
 pub const BEEFY_CLIENT_STATE_TYPE_URL: &str = "/ibc.lightclients.beefy.v1.ClientState";
 pub const MOCK_CLIENT_STATE_TYPE_URL: &str = "/ibc.mock.ClientState";
 
-pub trait ClientState: Clone + core::fmt::Debug + Send + Sync {
+pub trait ClientState: Clone + Debug + Send + Sync {
     /// Client-specific options for upgrading the client
     type UpgradeOptions;
 
+    /// Client definition type (used for verification)
+    type ClientDef: ClientDef;
+
     /// Return the chain identifier which this client is serving (i.e., the client is verifying
     /// consensus states from this chain).
     fn chain_id(&self) -> ChainId;
 
     /// Type of client associated with this state (eg. Tendermint)
     fn client_type(&self) -> ClientType;
 
+    /// Returns a client definition for this client state
+    fn client_def(&self) -> Self::ClientDef;
```

Comment:
You could remove this function, it doesn't seem to be used anywhere, that way `AnyClientState`,  `BeefyClientState` and `TendermintClientState` won't have to be generic anymore